sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala

```diff
@@ -160,12 +160,12 @@ class OrcFileFormat
     }
 
     val resultSchema = StructType(requiredSchema.fields ++ partitionSchema.fields)
+    val actualSchema = StructType(dataSchema.fields ++ partitionSchema.fields)
     val sqlConf = sparkSession.sessionState.conf
     val enableVectorizedReader = supportBatch(sparkSession, resultSchema)
     val capacity = sqlConf.orcVectorizedReaderBatchSize
 
-    val resultSchemaString = OrcUtils.orcTypeDescriptionString(resultSchema)
-    OrcConf.MAPRED_INPUT_SCHEMA.setString(hadoopConf, resultSchemaString)
```

Comment:
After reading the code more, I think this is the real problem. When the physical file schema doesn't match the table schema, this `resultSchemaString` is wrong.

I think we should set this config in the executor side, where we know the physical file schema. The `OrcUtils.requestedColumnIds` should not only report the column indices, but also the actual `requiredSchema`.

Reply:
will this not help 
I have already added this in the code
`OrcConf.MAPRED_INPUT_SCHEMA.setString(conf, resultSchemaString)`

Reply:
@cloud-fan - Do we need to make the similar change in these file OrcPartitionReaderFactory

override def buildReader(file: PartitionedFile): PartitionReader[InternalRow] = {
and 
override def buildColumnarReader(file: PartitionedFile): PartitionReader[ColumnarBatch] = {